modules/geochemistry/src/utils/EquilibriumGeochemicalSolver.C

```diff
@@ -0,0 +1,393 @@
+  // potentials
+  ind = 0;
+  std::vector<unsigned> res_order(_num_basis_in_algebraic_system);
+  std::iota(res_order.begin(), res_order.end(), ind++);
+  std::sort(res_order.begin(), res_order.end(), [&](int i, int j) {
+    return std::abs(_residual(i)) > std::abs(_residual(j));
+  });
+
+  const std::vector<Real> original_molality_and_pot = _egs.getAlgebraicVariableValues();
+  DenseVector<Real> new_molality_and_pot(original_molality_and_pot);
+  for (const auto & a : res_order)
+  {
+    if (std::abs(_residual(a)) < _max_initial_residual)
+      return false; // haven't managed to find a suitable new molality, and all remaining residual
+                    // components are less than the cutoff, so cannot appreciably reduce from now
+                    // on
+
+    const Real multiplier = (original_molality_and_pot[a] > median_molality) ? 0.5 : 2.0;
+    // try using the multiplier
+    new_molality_and_pot(a) = multiplier * original_molality_and_pot[a];
+    _egs.setAlgebraicVariables(new_molality_and_pot);
+    _abs_residual = computeResidual(_residual);
+    if (_abs_residual < initial_r)
+      return true; // success: found a new molality that decreases the initial |R|
+
+    // the above approach did not decrease |R|, so try using 1/multiplier
+    new_molality_and_pot(a) = original_molality_and_pot[a] / multiplier;
+    _egs.setAlgebraicVariables(new_molality_and_pot);
+    _abs_residual = computeResidual(_residual);
+    if (_abs_residual < initial_r)
+      return true; // success: found a new molality that decreases the initial |R|
+
+    // the new molalities did not decrease |R|, so revert to the original molality, and move to
+    // next-biggest residual component
+    new_molality_and_pot(a) = original_molality_and_pot[a];
+    _egs.setAlgebraicVariables(new_molality_and_pot);
+    _abs_residual = computeResidual(_residual);
+  }
+  return false;
+}
+
+void
+EquilibriumGeochemicalSolver::solveSystem(std::stringstream & ss,
+                                          unsigned & tot_iter,
+                                          Real & abs_residual)
+{
+  ss.str("");
+  tot_iter = 0;
+
+  bool still_swapping = true;
+  while (still_swapping)
+  {
+    _num_basis_in_algebraic_system = _egs.getNumBasisInAlgebraicSystem();
+    _num_in_algebraic_system = _egs.getNumInAlgebraicSystem();
+    _residual = DenseVector<Real>(_num_in_algebraic_system);
+    _jacobian = DenseMatrix<Real>(_num_in_algebraic_system, _num_in_algebraic_system);
+    _new_mol = DenseVector<Real>(_num_in_algebraic_system);
+
+    unsigned iter = 0;
+    const Real max_is0 =
+        std::min(1.0, (iter + 1.0) / (_ramp_max_ionic_strength + 1.0)) * _max_ionic_strength;
+    _is.setMaxIonicStrength(max_is0);
+    _is.setMaxStoichiometricIonicStrength(max_is0);
+    _abs_residual = computeResidual(_residual);
+    bool reducing_initial_molalities = (_abs_residual > _max_initial_residual);
+    while (reducing_initial_molalities)
```

Comment:
Is this guaranteed to converge, always? could be stuck in an infinite loop

Reply:
An infinite loop is impossible here, but a very long loop is possible.  i'll put in a counter to abort the loop if it goes on too long.